File: src/test/java/by/architecture/map/service/impl/SourceServiceImplTest.java
package by.architecture.map.service.impl;

import by.architecture.map.dto.SourceDto;
import by.architecture.map.entity.Source;
import by.architecture.map.exception.SourceException;
import by.architecture.map.mapper.SourceMapper;
import by.architecture.map.repository.SourceRepository;
import lombok.SneakyThrows;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.AdditionalAnswers.answer;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class SourceServiceImplTest {

    @Mock
    private SourceRepository sourceRepository;

    @Mock
    private SourceMapper sourceMapper;

    @InjectMocks
    private SourceServiceImpl sourceService;

    private SourceDto sourceDto;
    private Source source;
    private Integer id;

    @BeforeEach
    public void init() {

        sourceDto = SourceDto.builder()
                .name("My home is Belarus")
                .url("my-home-is-belarus.com")
                .build();

        id = 1;

        source = Source.builder()
                .id(id)
                .name(sourceDto.getName())
                .url(sourceDto.getUrl())
                .build();
    }

    @Test
    @SneakyThrows
    void givenSourceDto_whenSavingSourceDto_thenReturnSavedSource() {

        //Arrange
        when(sourceRepository.save(any(Source.class))).thenReturn(source);
        when(sourceMapper.toSource(any(SourceDto.class))).thenReturn(source);

        //Act
        Source actual = sourceService.add(sourceDto);

        //Assert
        assertThat(actual).isEqualTo(source);
    }

    @Test
    void givenSourceListWithElement_whenFindAllSources_thenGetSourceListSizeGreaterThanNull() {

        //Arrange
        List<Source> sources = new ArrayList<>();
        sources.add(source);
        when(sourceRepository.findAll()).thenReturn(sources);

        //Act
        List<Source> actual = sourceService.findAll();

        //Assert
        assertThat(actual.size()).isGreaterThan(0);
    }

//    @Test
//    void  update() throws SourceException {
//
//        //Arrange
//        SourceDto sourceUpdates = SourceDto.builder()
//                .name("Радзіма")
//                .url("radzima.by")
//                .build();
//
//        when(sourceMapper.toSource(sourceUpdates)).thenReturn(source);
////        doNothing().when(sourceRepository.findById(anyInt())
////                .orElseThrow(() -> new SourceException("Source with id = " + id + "not exists.")));
//
//        when(sourceRepository.findById(anyInt())
//                .orElseThrow(() -> new SourceException("Source with id = " + id + "not exists.")))
//                .thenReturn(source);
//
//        //Act
//        Source actual = sourceService.update(id, sourceUpdates);
//
//        //Assert
//        assertThat(source).isEqualTo(actual);
//    }


    @Test
    @SneakyThrows
    void givenSourceId_whenDeleteSource_thenDeleteSource() {

        //Arrange
        doNothing().when(sourceRepository).deleteById(id);
        when(sourceRepository.existsById(id)).thenReturn(true);

        //Act
        sourceService.delete(id);

        //Assert
        verify(sourceRepository, times(1)).deleteById(id);
    }

}

File: src/test/kotlin/by/architecturemap/belarus/service/impl/SourceServiceImplTest.kt
package by.architecturemap.belarus.service.impl

import by.architecturemap.belarus.dto.SourceDTO
import by.architecturemap.belarus.entity.Source
import by.architecturemap.belarus.exception.NotFoundException
import by.architecturemap.belarus.repository.jpa.SourceRepository
import io.mockk.Runs
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.assertThrows
import java.util.Optional

class SourceServiceImplTest {

    private val sourceRepository: SourceRepository = mockk()
    private val sourceService = SourceServiceImpl(sourceRepository)

    private val id = 1
    private lateinit var source: Source
    private lateinit var updatedSource: Source

    @BeforeEach
    fun setUp() {
        source = Source(name = "Source Name", url = "http://example.com", description = "Source Description")
        updatedSource = Source(name = "Updated Name", url = "http://new-url.com", description = "Updated Description")
    }

    @Test
    fun givenSource_whenCreate_thenReturnSavedSource() {
        // arrange
        val expected = source
        every { sourceRepository.save(source) } returns source

        // act
        val actual = sourceService.create(source)

        // assert
        verify(exactly = 1) { sourceRepository.save(source) }
        assertEquals(expected, actual)
    }

    @Test
    fun whenFindAll_thenReturnListOfSources() {
        // arrange
        val sources = listOf(source, updatedSource)
        every { sourceRepository.findAll() } returns sources

        // act
        val actual = sourceService.findAll()

        // assert
        verify(exactly = 1) { sourceRepository.findAll() }
        assertEquals(sources, actual)
    }

    @Test
    fun givenSourceId_whenUpdate_thenReturnUpdatedSource() {
        // arrange
        val expected = updatedSource
        every { sourceRepository.findById(id) } returns Optional.of(source)
        every { sourceRepository.save(source) } returns source

        // act
        val actual = sourceService.update(id, updatedSource)

        // assert
        verify(exactly = 1) { sourceRepository.findById(id) }
        verify(exactly = 1) { sourceRepository.save(source) }
        assertEquals(expected, actual)
    }

    @Test
    fun givenSourceDTO_whenPatchUpdate_thenReturnPatchedSource() {
        // arrange
        val sourceDTO = SourceDTO(
            name = "Updated Name",
            url = "http://example2.com",
            description = "Updated Description"
        )

        every { sourceRepository.findById(id) } returns Optional.of(source)
        every { sourceRepository.save(source) } returns source

        // act
        val actual = sourceService.patchUpdate(id, sourceDTO)

        // assert
        val expected = Source(
            name = "Updated Name",
            url = "http://example2.com",
            description = "Updated Description"
        )
        verify(exactly = 1) { sourceRepository.findById(id) }
        verify(exactly = 1) { sourceRepository.save(source) }
        assertEquals(expected, actual)
    }

    @Test
    fun givenBlankSourceDTO_whenPatchUpdate_thenReturnOriginalSource() {
        // arrange
        val sourceDTO = SourceDTO(
            name = "",
            url = "",
            description = ""
        )

        every { sourceRepository.findById(id) } returns Optional.of(source)
        every { sourceRepository.save(source) } returns source

        // act
        val actual = sourceService.patchUpdate(id, sourceDTO)

        // assert
        val expected = source
        verify(exactly = 1) { sourceRepository.findById(id) }
        verify(exactly = 1) { sourceRepository.save(source) }
        assertEquals(expected, actual)
    }

    @Test
    fun givenSourceDTOWithNullFields_whenPatchUpdate_thenReturnOriginalSource() {
        // arrange
        val sourceDTO = SourceDTO(
            name = null,
            url = null,
            description = null
        )

        every { sourceRepository.findById(id) } returns Optional.of(source)
        every { sourceRepository.save(source) } returns source

        // act
        val actual = sourceService.patchUpdate(id, sourceDTO)

        // assert
        val expected = source
        verify(exactly = 1) { sourceRepository.findById(id) }
        verify(exactly = 1) { sourceRepository.save(source) }
        assertEquals(expected, actual)
    }

    @Test
    fun givenSourceId_whenDelete_thenVerifyDeletion() {
        // arrange
        every { sourceRepository.findById(id) } returns Optional.of(source)
        every { sourceRepository.deleteById(id) } just Runs

        // act
        sourceService.delete(id)

        // assert
        verify(exactly = 1) { sourceRepository.findById(id) }
        verify(exactly = 1) { sourceRepository.deleteById(id) }
    }

    @Test
    fun givenInvalidSourceId_whenDelete_thenThrowNotFoundException() {
        // arrange
        every { sourceRepository.findById(id) } returns Optional.empty()

        // act & assert
        assertThrows<NotFoundException> { sourceService.delete(id) }

        // verify
        verify(exactly = 1) { sourceRepository.findById(id) }
        verify(exactly = 0) { sourceRepository.deleteById(id) }
    }
}
